### DISCUSSION TITLE
Why `code-server --help` so different in `macos` and `unbuntu`?

### DISCUSSION THREAD

**Discussion by qupig** [coder/code-server#7050] [General] [answered]
In both `macos` and `unbuntu`, use the following command to install directly:
```sh
curl -fsSL https://code-server.dev/install.sh | sh
```

Then, in macOS:
```log
$ code-server --help
code-server 4.93.1 69df01185ce2f80e99c9e4f8c8de1907cc7a9bc5 with Code 1.93.1

Usage: code-server [options] [path]
    - Opening a directory: code-server ./path/to/your/project
    - Opening a saved workspace: code-server ./path/to/your/project.code-workspace

Options
      --auth                             The type of authentication to use. [password, none]
      --password                         The password for password authentication (can only be passed in via $PASSWORD or the config file).
      --hashed-password                  The password hashed with argon2 for password authentication (can only be passed in via $HASHED_PASSWORD or the config file).
                                         Takes precedence over 'password'.
      --cert                             Path to certificate. A self signed certificate is generated if none is provided.
      --cert-host                        Hostname to use when generating a self signed certificate.
      --cert-key                         Path to certificate key when using non-generated cert.
      --disable-telemetry                Disable telemetry.
      --disable-update-check             Disable update check. Without this flag, code-server checks every 6 hours against the latest github release and
                                         then notifies you once every week that a new release is available.
      --disable-file-downloads           Disable file downloads from Code. This can also be set with CS_DISABLE_FILE_DOWNLOADS set to 'true' or '1'.
      --disable-file-uploads             Disable file uploads.
      --disable-workspace-trust          Disable Workspace Trust feature. This switch only affects the current session.
      --disable-getting-started-override Disable the coder/coder override in the Help: Getting Started page.
      --disable-proxy                    Disable domain and path proxy routes.
   -h --help                             Show this output.
      --locale                           Set vscode display language and language to show on the login page, more info see
                                         https://en.wikipedia.org/wiki/IETF_language_tag
      --open                             Open in browser on startup. Does not work remotely.
      --bind-addr                        Address to bind to in host:port. You can also use $PORT to override the port.
      --config                           Path to yaml config file. Every flag maps directly to a key in the config file.
      --socket                           Path to a socket (bind-addr will be ignored).
      --socket-mode                      File mode of the socket.
      --trusted-origins                  Disables authenticate origin check for trusted origin. Useful if not able to access reverse proxy configuration.
   -v --version                          Display version information.
      --user-data-dir                    Path to the user data directory.
      --extensions-dir                   Path to the extensions directory.
      --list-extensions                  List installed VS Code extensions.
      --force                            Avoid prompts when installing VS Code extensions.
      --install-extension                Install or update a VS Code extension by id or vsix. The identifier of an extension is `${publisher}.${name}`.
                                         To install a specific version provide `@${version}`. For example: 'vscode.csharp@1.2.3'.
      --enable-proposed-api              Enable proposed API features for extensions. Can receive one or more extension IDs to enable individually.
      --uninstall-extension              Uninstall a VS Code extension by id.
      --show-versions                    Show VS Code extension versions.
      --github-auth                      GitHub authentication token (can only be passed in via $GITHUB_TOKEN or the config file).
      --proxy-domain                     Domain used for proxying ports.
   -e --ignore-last-opened               Ignore the last opened directory or workspace in favor of an empty window.
   -n --new-window                       Force to open a new window.
   -r --reuse-window                     Force to open a file or folder in an already opened window.
 -vvv --verbose                          Enable verbose logging.
  -an --app-name                         The name to use in branding. Will be shown in titlebar and welcome message
   -w --welcome-text                     Text to show on login page
      --abs-proxy-base-path              The base path to prefix to all absproxy requests
```

In Ubuntu:
```
$ code-server --help
code-server 1.93.1

Usage: code-server [options][paths...]

To read from stdin, append '-' (e.g. 'ps aux | grep code | code-server -')

Options
  -d --diff <file> <file>                    Compare two files with each other.
  -m --merge <path1> <path2> <base> <result> Perform a three-way merge by providing paths for two modified versions of a file, the
                                             common origin of both modified versions and the output file to save merge results.
  -a --add <folder>                          Add folder(s) to the last active window.
  -g --goto <file:line[:character]>          Open a file at the path on the specified line and character position.
  -n --new-window                            Force to open a new window.
  -r --reuse-window                          Force to open a file or folder in an already opened window.
  -w --wait                                  Wait for the files to be closed before returning.
  -h --help                                  Print usage.
  -c --stdin-to-clipboard                    copies the STDIN to the clipboard

Extensions Management
  --list-extensions                   List the installed extensions.
  --show-versions                     Show versions of installed extensions, when using --list-extensions.
  --category <category>               Filters installed extensions by provided category, when using --list-extensions.
  --install-extension <ext-id | path> Installs or updates an extension. The argument is either an extension id or a path to a
                                      VSIX. The identifier of an extension is '${publisher}.${name}'. Use '--force' argument to
                                      update to latest version. To install a specific version provide '@${version}'. For example:
                                      'vscode.csharp@1.2.3'.
  --uninstall-extension <ext-id>      Uninstalls an extension.
  --update-extensions                 Update the installed extensions.

Troubleshooting
  -v --version Print version.
  --verbose    Print verbose output (implies --wait).
  -s --status  Print process usage and diagnostics information.
```

**Comment by lg-nineoneninethreenine** [comment]
code-server on macOS and Ubuntu are running on different versions - check the output you'll see it. Basically Ubuntu is running a very old version of code-server.
There's something off with this, as the cURL command installs the latest version of code-server for the system, along with the specific architecture(amd64, arm64, etc.). Re-run the cURL command, but dry run it, and spot which version it mentions.

Edit: the version changed from 4.93.1(macOS) to 1.93.1(Ubuntu Linux). Something _may_ be wrong with install.sh

**Reply by lg-nineoneninethreenine** [reply]
Update: I tested this on a cloud instance of Ubuntu(Gitpod), and everything looks good. @qupig can you tell me which version of Ubuntu you may be using? Basically info about the Ubuntu system/VM/instance.
It might be due to something in the system, since the Ubuntu instance I just used is 22.04, and the dry-run looks good.
https://asciinema.org/a/UMcTWzWj5ag8LZSnNkGekQ6vB
The link above is an asciinema recording of what I re-ran for proof.

**Reply by qupig** [reply]
Looking at the installation output, I don't think an older version is installed?

```log
> curl -fsSL https://code-server.dev/install.sh | sh
Ubuntu 22.04.5 LTS
Installing v4.93.1 of the amd64 deb package from GitHub.
+ mkdir -p ~/.cache/code-server
+ curl -#fL -o ~/.cache/code-server/code-server_4.93.1_amd64.deb.incomplete -C - https://github.com/coder/code-server/releases/download/v4.93.1/code-server_4.93.1_amd64.deb
######################################################################## 100.0%
+ mv ~/.cache/code-server/code-server_4.93.1_amd64.deb.incomplete ~/.cache/code-server/code-server_4.93.1_amd64.deb
+ sudo dpkg -i ~/.cache/code-server/code-server_4.93.1_amd64.deb
Selecting previously unselected package code-server.
(Reading database ... 266916 files and directories currently installed.)
Preparing to unpack .../code-server_4.93.1_amd64.deb ...
Unpacking code-server (4.93.1) ...
Setting up code-server (4.93.1) ...
deb package has been installed.
```

**Reply by lg-nineoneninethreenine** [reply]
Huh...
Well that's odd, can you check if you have the latest version now?
Edit: this isn't from the asciinema which then means that you have the latest LTS Ubuntu version. But then this isn't possible, you can't have an older version.

**Comment by qupig** [comment]
I think I've found the cause of the issue, but I'm not sure how it happened.

```log
> /usr/bin/code-server -v
4.93.1 69df01185ce2f80e99c9e4f8c8de1907cc7a9bc5 with Code 1.93.1

> /usr/lib/code-server/bin/code-server -v
4.93.1 69df01185ce2f80e99c9e4f8c8de1907cc7a9bc5 with Code 1.93.1

> which code-server
/usr/lib/code-server/lib/vscode/bin/remote-cli/code-server

> /usr/lib/code-server/lib/vscode/bin/remote-cli/code-server -v
1.93.1
3542bd157b0ccc8bf732acdb03d4d69d85f48ea9
x64
```

Again, I don't know why:
```log
> echo $PATH
/usr/lib/code-server/lib/vscode/bin/remote-cli:...:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

I am running the above commands in `code-server` instance, that is vscode’s built-in `TERMINAL` panel with `bash`. 
Do you think it's possible that in the instance,`/usr/lib/code-server/lib/vscode/bin/remote-cli` was automatically added to `$PATH`?

Anyway, the same way of doing things, which works without issue in `macOS`, but occurs in `Ubuntu`.

**Reply by lg-nineoneninethreenine** [reply]
Hold on a minute, the wrong directory has a greater priority than the correct directory!
Tryh this @qupig: alter the PATH variable such that the remote CLI directory(`/usr/lib/code-server/lib/vscode/bin/remote-cli`) is at the end than at the very beginning, and restart your shell. Then run code-server, let's see if this fixes it.
Don't remove anything, just rearrange it, and only one thing should be moved(`/usr/lib/code-server/lib/vscode/bin/remote-cli`).

**Reply by lg-nineoneninethreenine** [reply]
If you want, for safety, check your PATH environmental variable and check which ones need more priority, like system directories above installed directories. This could help solve this issue fully.

**Reply by qupig** [reply]
There is no need for this test, that much is obvious. 
The crux of the matter is why the path was added to the environment variable. 
Who added it and when. 
Like I said, my guess is that the `code-server` instance added it.
The issue here is that the Coder’s cli wrapper and the cli in `vscode/bin/remote-cli` have the same name, and the latter was added to `$PATH` at some point.

**Reply by lg-nineoneninethreenine** [reply]
This is just a theory, but a very probable one. You added code-server to PATH after installing, but I think a mishap might have happened, since it seems like instead of appending the directory to PATH, it appended the existing PATH value to the directory.
It wasn't the case with my Gitpod test maybe because I didn't add code-server to PATH. I think you might have done that

**Reply by qupig** [reply]
@im-coder-lg I have never manually modified `PATH`, everything is done automatically by `install.sh` or `code-server` instance.

**Reply by qupig** [reply]
I tested it on a fresh Ubuntu system and my suspicions were confirmed. 
The `/usr/lib/code-server/lib/vscode/bin/remote-cli` is added to `PATH` to the built-in `TERMINAL` by the `code-server` instance.
That is, in the external SHELL environment, there is no such extra path, and that path only exists in the terminal in the `code-server` instance.

**Reply by lg-nineoneninethreenine** [reply]
Well then you can fix it right? Something might be wrong with code-server here, so once I test it I'll ping maintainers and let them know. @qupig can you try fixing this issue by editing the PATH variable? Or can you edit it in the first place?

**Reply by qupig** [reply]
Why should I edit `PATH` env?
Apparently editing the `PATH` env could fixes it, but I didn't add it and I shouldn't edit or delete it.
I could directly run `/usr/bin/code-server --help` instead.
There are already so many identical [issues](https://github.com/coder/code-server/discussions/7050#discussioncomment-11057493) out there that you can definitely report it to the maintainers if they are really willing to fix it.

**Reply by lg-nineoneninethreenine** [reply]
You could add an alias to your shell or try creating a symbolic link, something might help remedy that. I can't imagine anyone running code-server by putting the actual executable in, so that oughta remedy your issue for now.
I'll try something and reproduce this, and maybe I'll get some idea on a fix.

**Reply by lg-nineoneninethreenine** [reply]
Unfortunately, I cannot recreate this issue, maybe due to factors with Gitpod.
I looked up the other issues, and I understand one thing: Microsoft's code-server gets more priority than our code-server. I think Microsoft is issuing PATH as `[path-to-their-code-server]:$PATH` but they aren't fixing that, and that's what is leading to this conflict. Moreover, both programs share the same name. There is no permanent fix for this yet, but rearranging PATH, creating an alias, or trying some symbolic link might work.

**Comment by qupig** [comment, marked as answer]
Actually:
- https://github.com/coder/code-server/issues/5335
- https://github.com/coder/code-server/pull/5360
- https://github.com/coder/code-server/issues/5502
- https://github.com/coder/code-server/issues/5621#issuecomment-1268069226
- https://github.com/coder/code-server/issues/6240#issuecomment-1607308944

**Reply by qupig** [reply]
The issue is years old, it doesn't seem like it will be fixed, and it's not well documented. 
This means that everyone falls down this rabbit hole. 😫

**Comment by code-asher** [comment]
If I understand correctly, what is going on here is that in some cases our `code-server` command is being ran, and in other cases VS Code's `code-server` is being ran.

VS Code's runs in the integrated terminal by default (it gets automatically added to `PATH`).  This version of `code-server` allows interacting with the current instance of code-server (for example opening a file in the current instance), but it does not allow spawning a full new code-server instance; they are different scripts and have different options and purposes (although there is overlap).  

Understandably confusing!  This is how it works upstream though, so it is not specific to code-server, and appears to be the intended behavior.

**Reply by qupig** [reply]
Like you said:
- Inconsistencies are confusing
- Inconsistent behavior in macOS/Ubuntu (Even more confused)

**Reply by code-asher** [reply]
Yeah I agree.  It does seem odd that the behavior is different in macOS and Ubuntu though, that could be a bug.  Is your mac also using Bash?  And does it have terminal shell integration enabled in the settings?

**Reply by code-asher** [reply]
Ah I think I found where they do it, still not sure why it might not be taking effect in macOS though: https://github.com/microsoft/vscode/blob/release/1.95/src/vs/server/node/extensionHostConnection.ts#L54-L60

**Reply by qupig** [reply]
> Yeah I agree. It does seem odd that the behavior is different in macOS and Ubuntu though, that could be a bug. Is your mac also using Bash? And does it have terminal shell integration enabled in the settings?

Yes it is `bash`, I'm not sure where the setting you are talking about is?
If you need me to do any tests or check anything please let me know.

**Reply by code-asher** [reply]
This setting here (`terminal.integrated.shellIntegration.enabled`):

![screenshot](https://github.com/user-attachments/assets/2e67e8fc-1ab9-42ed-8e81-41c383519397)

Also could you check whether `terminal.integrated.inheritEnv` is checked?

I am not very familiar with this code, so kind of just guessing.

**Reply by qupig** [reply]
@code-asher 
`terminal.integrated.shellIntegration.enabled` = true
`terminal.integrated.inheritEnv` = true

And I want to say that I didn't change any vscode settings, in the test environment.

And as you can see, the issue comes from the difference in PATH order:

macOS-PATH
```
/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/local/sbin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/opt/homebrew/Cellar/code-server/4.93.1/libexec/lib/node_modules/code-server/lib/vscode/bin/remote-cli:/opt/homebrew/lib/ruby/gems/3.0.0/bin:/opt/homebrew/opt/ruby@3.0/bin
```

Ubuntu-PATH
```
/usr/lib/code-server/lib/vscode/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Reply by code-asher** [reply]
Ahhh I see, so the `remote-cli` path is being prepended, but then other things get prepended as well.

I would guess something in a `.bashrc` or similiar file is causing this, but it could also be a bug in VS Code.

**Reply by qupig** [reply]
I investigated further and discovered something interesting.

Creating a new terminal in vscode in Ubuntu will not reload files such as `/etc/profile`, `~/.bash_profile`, but `~/.bashrc` will.

But in macOS, all these files are reloaded in each new terminal instance.

In macOS, at the beginning of `/etc/profile`, `remote-cli` is indeed first in `PATH`, but is changed by the first piece of code that follows:
```sh
if [ -x /usr/libexec/path_helper ]; then
	eval `/usr/libexec/path_helper -s`
fi
```

I think this explains the difference results in the two systems. 

But why the behavior of reading shell configuration files is different in the two systems is still unclear.

**Reply by code-asher** [reply]
Good find!  `/etc/profile` is only sourced if the shell is interactive or ran with `--login`, I think?  Maybe there is something about that which explains the different behavior, like VS Code is using a login shell in one case but not the other, for some reason.

**Reply by qupig** [reply]
https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html
https://www.gnu.org/software/bash/manual/html_node/The-Shopt-Builtin.html
https://www.gnu.org/software/bash/manual/html_node/Is-this-Shell-Interactive_003f.html

But based on simple inspection, it seems that both are non-login interactive shell:

macOS:
```
> shopt login_shell
login_shell     off
> logout
bash: logout: not login shell: use `exit'
> echo $-
himBH
```

Ubuntu:
```
> shopt login_shell
login_shell     off
> logout
bash: logout: not login shell: use `exit'
> echo $-
himBHs
```

**Reply by qupig** [reply]
Possibly related:
https://github.com/microsoft/vscode/blob/db7e872d7be8f0a5e01316d4295cbb2c92583d29/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L15-L39

macOS:
```
bash --init-file /opt/homebrew/Cellar/code-server/4.93.1/libexec/lib/node_modules/code-server/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
```

Ubuntu:
```
bash --init-file /usr/lib/code-server/lib/vscode/out/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh
```

**Reply by code-asher** [reply]
Interesting.  When I download the macOS tar.gz for 4.95.1, there is no `lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh` file, only the one in `contrib/terminal`.

Maybe that only means the file moved, but maybe it also means they changed something in this area and updating to 4.95.1 will have different behavior?

**Reply by code-asher** [reply]
Or wait, it is sourcing `/etc/profile` in both macOS and Ubuntu right?  But the difference is simply that in macOS your `/etc/profile` has that line changing `PATH` where Ubuntu has no such thing.

Ah but you said Ubuntu is not reading `/etc/profile`.  Looks like it should, based on the code, though, if `VSCODE_SHELL_LOGIN` is set.  :thinking:

**Reply by code-asher** [reply]
This, maybe? https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/terminal/node/terminalEnvironment.ts#L188-L194

**Reply by qupig** [reply]
Yes, it looks like the file path has changed:

```
/code-server-4.95.1-macos-arm64.tar/code-server-4.95.1-macos-arm64/lib/vscode/out/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh
/code-server-4.93.1-macos-arm64.tar/code-server-4.93.1-macos-arm64/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
```

> Or wait, it is sourcing /etc/profile in both macOS and Ubuntu right?
> Ah but you said Ubuntu is not reading `/etc/profile`

Yes. From the behavior I've observed, only macOS reads the file when creating a new terminal panel.
I confirmed this by adding an `echo` line to the beginning of these files.
So the strange thing is why `non-login` shells in macOS read these files, `/etc/profile`, `~/.bash_profile` etc.
I tend to think that is handled internally by vscode. But it could also be other reasons that I haven't thought of.
The above `shellIntegration-bash.sh` script was found by inspect the current terminal through `pstree`, like (macos):
```
> pstree -p $$
-+= 00001 root /sbin/launchd
 \-+- 06560 user /opt/homebrew/opt/node@20/bin/node /opt/homebrew/bin/code-server
   \-+- 06561 user /opt/homebrew/Cellar/node@20/20.18.0_2/bin/node /opt/homebrew/Cellar/code-server/4.93.1/libexec/lib/node_modules/code-server/out/node/entry
     \-+- 06884 user /opt/homebrew/Cellar/node@20/20.18.0_2/bin/node /opt/homebrew/Cellar/code-server/4.93.1/libexec/lib/node_modules/code-server/lib/vscode/out/bootstrap-fork --type=ptyHost --logsPath /Users/user/.local/share/code-server/logs/202411
       \-+= 06885 user /bin/bash --init-file /opt/homebrew/Cellar/code-server/4.93.1/libexec/lib/node_modules/code-server/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
         \-+= 07295 user pstree -p 6885
           \--- 07296 root ps -axwwo user,pid,ppid,pgid,command
```

> This, maybe?

I have seen this file by searching for `VSCODE_SHELL_LOGIN`, but I don't know its internal workings.

**Reply by code-asher** [reply]
Yeah I am not sure how it works either.  I added `echo` to `/etc/profile` on Ubuntu and I do not see it logged either.  I do not have a macOS device to test though.

What if you start `code-server` with `--log trace` and compare the output between Ubuntu and macOS?  For example based on the links we posted, a login shell is only used if `VSCODE_SHELL_LOGIN` is set, and it is not set if there are no args, and I see no args here in my logs:

```
shellLaunchConfig: {
    name: undefined,
    executable: '/nix/var/nix/profiles/default/bin/bash',
    args: undefined,
    cwd: '/home/coder/coder',
    env: undefined,
    useShellEnvironment: true,
    reconnectionProperties: undefined,
    type: undefined,
    isFeatureTerminal: undefined
  },
```

So my guess is that on macOS there will be `args` here, and if so the next question will be why.

**Reply by code-asher** [reply]
Maybe this has a clue: https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/shell/node/shellEnv.ts#L131-L151

**Reply by code-asher** [reply]
Hmm or maybe not, I see in my logs `getUnixShellEnvironment#spawn ["-i","-l","-c"]`.  I guess these arguments are dropped at some point?

Oh...actually looks like they spawn a shell with `-i -l -c` to get the environment variables, and this is completely separate from the shell spawned for the terminal window in VS Code.

Ah so this could explain it actually.  They run this, the `PATH` gets set, but the `echo` does not show because this is a separate thing run only to get the env.  Then they run another shell without any login arguments for the actual terminal session the user will see.

**Reply by code-asher** [reply]
Yeah if I add `touch /tmp/bash-log` to my `/etc/profile` and restart code-server, that file is created, so `/etc/profile` is being ran at least on startup.

**Reply by qupig** [reply]
Great find, I really wasn't expecting this. Yes, `touch` is a smarter way than `echo` to stdout.

**Reply by qupig** [reply]
Actually we're a little off track, it's not necessary to investigate the `code-server` startup process.

Because the fact is that before starting `code-server`, the `PATH` in the system terminal does not have the `remote-cli` PATH, which is added to the built-in terminal by `vscode` after starting.

I tried `touch` and directly modify the `PATH` in Ubuntu's `/etc/profile`, and obviously it was not loaded in the built-in new terminal panel.

So this doesn't change my previous question of why vscode in macOS sourced those configuration files in the built-in terminal.

In other words, `remote-cli` is appended to the beginning of `PATH` in both systems, but because `/etc/profile` is re-sourced in macOS, and that `path_helper` changes the PATH order again, resulting in different results.

So we don't need to pay attention to the startup process of `code-server`. Instead, focus on the initialization process of the built-in terminal panel.

I've tried enable in vscode and viewing the trace log like they [mentioned](https://github.com/microsoft/vscode/issues/99878#issuecomment-642733043), but I found no information of note and there were actually too many log entries to make sense of. (In fact, the issue itself may be the clue and reason for the inconsistent behavior of macOS, it's [mentioned](https://github.com/microsoft/vscode/blob/db7e872d7be8f0a5e01316d4295cbb2c92583d29/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L35) in the code block I linked to earlier)

**Reply by code-asher** [reply]
Ah see I was under the impression that VS Code runs this on startup to get the environment variables, and then injects the environment variables that it fetched during startup into each subsequent terminal.  So in that sense the startup affects the init of each terminal process.

So `/etc/profile` is sourced in both cases, and affects every terminal, but only macOS's `/etc/profile` is munging `PATH`, is how it looks to me at the moment.

**Reply by qupig** [reply]
> but only macOS's `/etc/profile` is munging PATH

This is not the crux of the matter. Because when prints PATH in `/etc/profile` first line in macOS, `remote-cli` is indeed ordered first, so obviously when starting `code-server`, vscode adds `remote-cli` after `/etc/profile` sourced. So the behavior of macOS does not make a difference.

The problem is that when creating each new terminal panel in vscode in macOS, the `/etc/profile` file will be sourced again, which is the reason for the difference.

And vscode will also source `~/.bash_profile` every time in new terminal panel in macOS. In other words, even if I have a line `PATH=somenew:$PATH` in `~/.bash_profile`, it will have an impact on the new terminal panel.

In Ubuntu, vscode will not source the `/etc/profile` file again when creating a new terminal panel. Even if I modify the PATH in `/etc/profile` after `code-server` running, it will not have any impact.

So the problem is not with macOS's behavior in `/etc/profile`.
It's the different behavior of `VSCode` in `macOS` and `Ubuntu`.

**Reply by code-asher** [reply]
> Because when prints PATH in /etc/profile first line in macOS, remote-cli is indeed ordered first, so obviously when starting code-server, vscode adds remote-cli after /etc/profile sourced.

Ahhhhh I think I understand now, when you say printing the first line in `/etc/profile` shows `remote-cli`, you mean when it is sourced for each terminal, not when it is sourced at startup.

1. VS Code sources `/etc/profile` on startup (no `remote-cli` in `PATH` yet)
2. VS Code add `remote-cli` to `PATH`
3. On macOS only, VS Code sources `/etc/profile` again on every new terminal (`remote-cli` is already in PATH)

Quite the mystery.  Would you happen to have a `~/.bashrc` that sources `/etc/profile`?  Maybe it is coming from VS Code, but maybe something else is sourcing it.

**Reply by qupig** [reply]
> you mean when it is sourced for each terminal, not when it is sourced at startup.

Yes, sorry for not highlighting this before.

> Would you happen to have a `~/.bashrc` that sources `/etc/profile`

No, I checked this. I also checked their loading order,

**Reply by code-asher** [reply]
Ah wait this could be it? https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts#L248-L250

**Reply by qupig** [reply]
I don't know, but don't forget I [checked before](https://github.com/coder/code-server/discussions/7050#discussioncomment-11171653) that it's not `login-shell`.

**Reply by code-asher** [reply]
Ah yeah, I just remembered that.  Also I think this is used for tasks and debug, not terminal.

**Reply by code-asher** [reply]
OK, pretty sure I found it this time: https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/terminal/common/terminalPlatformConfiguration.ts#L229-L255

What happens:

- OSX [default profile](https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/terminal/common/terminalPlatformConfiguration.ts#L236) has `-l` in it.
- VS Code [sees](https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/terminal/node/terminalEnvironment.ts#L190-L191) the `-l` and sets `VSCODE_SHELL_LOGIN=1`
- It [drops](https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/platform/terminal/node/terminalEnvironment.ts#L193) the `-l` when spawning `bash` and uses `bash --init-file shellIntegration-bash.sh` instead.
- The `VSCODE_SHELL_LOGIN` env var then [causes](https://github.com/microsoft/vscode/blob/65edc4939843c90c34d61f4ce11704f09d3e5cb6/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L22) `/etc/profile` and others to be sourced.

So it seems like this is intentional behavior.  I checked the commit and it looks like `-l` was added to fix this issue: https://github.com/microsoft/vscode/issues/124409

What they should probably do is prepend to the `PATH` after sourcing each time, instead of doing it once during startup.

It could be worked around by setting the profile:

```
"terminal.integrated.profile.osx": { "bash": null }
```

But then you might be missing a lot of other things from the path?  Not sure, because seems like it should already have the right PATH when it sources at startup, assuming the startup source does happen on macOS.

**Reply by qupig** [reply]
Great investigative report, this almost confirms our previous suspicions.

> It could be worked around by setting the profile

Tested this, no change. (I did not restart `code-server`)

**Reply by code-asher** [reply]
Strange, it worked for me!  (Tested it on a macOS VM.)  I do not think I restarted code-server either.  :thinking:

**Reply by code-asher** [reply]
Ah sorry I had a typo.  Could not figure out how to copy the value out of the VM so I tried to type it by hand.  Should be this, unless I made another typo :sweat_smile: : 

```
"terminal.integrated.profiles.osx": { "bash": null }
```

(`profiles` instead of `profile`)

**Reply by qupig** [reply]
It's really strange, I just tested it again and still no change.
The built-in terminal panel still suorce `/etc/profile` file.

And I found and copied the setting string from vscode settings so there won't be another typo. 😵‍💫

**Reply by qupig** [reply]
```json
"terminal.integrated.profiles.osx": {
    "bash": {
        "path": "bash",
        "args": null,
    },
}
```
Works for me, it no longer suorce `/etc/profile` file and the shell prompt is no longer stylized, and `remote-cli` is now first.

**Reply by code-asher** [reply]
Interesting that you have to specify `path` and `args` as well!  I am still not sure if the VS Code folks would consider this a bug but I wonder if we should open an issue upstream.  They should really add `remote-cli` after all the sourcing is done and before launching each terminal.

**Reply by qupig** [reply]
Not `path` and `args`, only `args` is the key, i think `path` may be omitted, didn't test, but I did omitted another parameter.

The problem is that it doesn't seem to override the default if set to `{ "bash": null }`, so there must be an object.

Unless they really want to maintain some compatibility with the project, there probably won't be any issues even if `remote-cli` is sorted later in the PATH without the project's `code-server` tool of the same name.

It's a bad things that two projects use the same command line tool name.

The inconsistency caused by the different PATH priorities of the two tools in the two systems is further confusion.

**Reply by code-asher** [reply]
Hmm yeah, I am not really sure what the best path forward is.  The problem is, if we replace theirs with ours, now we have to reimplement everything theirs has (like `--wait` for example is something we do not have), and if they ever change or add something, we have to do the same.  This is probably the best solution, but unfortunate.

**Reply by qupig** [reply]
I think the best way is to solve it fundamentally, that is, the two should use different cli tool names, which means that one of them needs to be renamed. 
It will be painful, but it will be the best for the future. 
To reduce the impact, a long migration period and good deprecation warnings and documentation are needed.

**Reply by code-asher** [reply]
You might be right, but we should do it without being a breaking change, which means we would have to implement all the missing flags into ours.

**Reply by code-asher** [reply]
Or actually maybe it would be OK if we said `to use this, run this command instead: <print the new name here>`

**Reply by qupig** [reply]
I recommend not mixing the flags of the two tools together, since that is actually maintained by two different teams, any future modifications will easily create new conflicts.

As a relatively easy alternative, and something that can be done by this downstream project, is to ensure that the project's `code-server` command always has priority in the PATH, whether it is an external terminal or an internal integrated terminal. (i.e. make sure it always overrides the upstream project and will not confuse users)

For the `code-server` command in the upstream original remote-cli, you could:

- Make an alias like `code-server-instance`
- It is detected that the `code-server` command is running in the inherited terminal, and a prompt is given at the beginning: <If you need to control this instance, please run the absolute path "/.../bin/remote-cli/code-server"> or if alias in last step <If you need to control this instance, please use `code-server-instance` command>

> Or actually maybe it would be OK if we said to use this, run this command instead: <print the new name here>

I recommend that you never do this. It prompts you to run another command after entering some flags exist.
This is definitely a disaster. Very bad user experience.
If you don't function/maintain these flags, don't add them.

**Reply by code-asher** [reply]
I agree giving ours priority is a good idea, but we have to implement the missing flags first (`--wait`, etc) to avoid a breaking change.

Unfortunately, I think we are stuck copying the flags in order to support them in external terminals.

**Reply by qupig** [reply]
Suffice it to say, if "breaking change" ends the ongoing chaos that has been and will continue to be, then it's worth it.
Regardless, at least having this documented well in somewhere prevents others from suffering the same confusion.
Thank you for your work and I will respect your decision.

**Reply by code-asher** [reply]
I appreciate your input!

Copying the flags is pretty quick so if we did this I would like to copy the missing ones real quick first and skip the breaking change.

Truthfully though, we update VS Code and merge PRs but are not actively working on the project so it may not happen for a while 😅
